File: server/tests/programs/javascript/05_complex_classes.js
// Complex Classes and OOP - Complexity: 5/6
// Tests ES6 classes, inheritance, and object-oriented patterns

// Base class
class Animal {
    constructor(name, age) {
        this.name = name;
        this.age = age;
    }
    
    speak() {
        console.log(`${this.name} makes a sound`);
    }
    
    getInfo() {
        return `${this.name} is ${this.age} years old`;
    }
}

// Inheritance
class Dog extends Animal {
    constructor(name, age, breed) {
        super(name, age);
        this.breed = breed;
    }
    
    speak() {
        console.log(`${this.name} barks: Woof!`);
    }
    
    fetch() {
        console.log(`${this.name} fetches the ball`);
    }
}

class Cat extends Animal {
    constructor(name, age, indoor) {
        super(name, age);
        this.indoor = indoor;
    }
    
    speak() {
        console.log(`${this.name} meows: Meow!`);
    }
    
    climb() {
        console.log(`${this.name} climbs a tree`);
    }
}

// Static methods and properties
class MathUtils {
    static PI = 3.14159;
    
    static circleArea(radius) {
        return this.PI * radius * radius;
    }
    
    static factorial(n) {
        if (n <= 1) return 1;
        return n * this.factorial(n - 1);
    }
}

// Getters and setters
class Rectangle {
    constructor(width, height) {
        this._width = width;
        this._height = height;
    }
    
    get area() {
        return this._width * this._height;
    }
    
    get perimeter() {
        return 2 * (this._width + this._height);
    }
    
    set width(value) {
        if (value > 0) this._width = value;
    }
    
    set height(value) {
        if (value > 0) this._height = value;
    }
}

// Test animals
console.log("=== Animal Classes ===");
const dog = new Dog("Buddy", 5, "Golden Retriever");
const cat = new Cat("Whiskers", 3, true);

console.log(dog.getInfo());
dog.speak();
dog.fetch();

console.log("\n" + cat.getInfo());
cat.speak();
cat.climb();

// Test static methods
console.log("\n=== Static Methods ===");
console.log(`Circle area (radius 5): ${MathUtils.circleArea(5).toFixed(2)}`);
console.log(`Factorial of 6: ${MathUtils.factorial(6)}`);

// Test getters/setters
console.log("\n=== Rectangle with Getters/Setters ===");
const rect = new Rectangle(10, 5);
console.log(`Dimensions: ${rect._width} x ${rect._height}`);
console.log(`Area: ${rect.area}`);
console.log(`Perimeter: ${rect.perimeter}`);

rect.width = 15;
console.log(`\nAfter width change to 15:`);
console.log(`New area: ${rect.area}`);
console.log(`New perimeter: ${rect.perimeter}`);

// Factory pattern
class ShapeFactory {
    static createShape(type, ...dimensions) {
        if (type === 'rectangle') {
            return new Rectangle(...dimensions);
        }
        return null;
    }
}

console.log("\n=== Factory Pattern ===");
const shape = ShapeFactory.createShape('rectangle', 7, 3);
console.log(`Created rectangle: Area = ${shape.area}`);

// Composition
class Engine {
    start() {
        return "Engine started";
    }
}

class Car {
    constructor(brand) {
        this.brand = brand;
        this.engine = new Engine();
    }
    
    start() {
        console.log(`${this.brand}: ${this.engine.start()}`);
    }
}

console.log("\n=== Composition Pattern ===");
const car = new Car("Tesla");
car.start();

File: server/tests/programs/cpp/05_complex_classes.cpp
// Complex Classes and OOP - Complexity: 5/6
// Tests class hierarchies, inheritance, and polymorphism

#include <iostream>
#include <string>
#include <cmath>
using namespace std;

// Base class
class Shape {
protected:
    string color;
    
public:
    Shape(string c) : color(c) {}
    
    virtual double area() const = 0;  // Pure virtual function
    virtual void display() const {
        cout << "Color: " << color << endl;
    }
    
    virtual ~Shape() {}  // Virtual destructor
};

// Derived class - Circle
class Circle : public Shape {
private:
    double radius;
    
public:
    Circle(string c, double r) : Shape(c), radius(r) {}
    
    double area() const override {
        return 3.14159 * radius * radius;
    }
    
    void display() const override {
        cout << "Circle - ";
        Shape::display();
        cout << "Radius: " << radius << ", Area: " << area() << endl;
    }
};

// Derived class - Rectangle
class Rectangle : public Shape {
private:
    double width, height;
    
public:
    Rectangle(string c, double w, double h) : Shape(c), width(w), height(h) {}
    
    double area() const override {
        return width * height;
    }
    
    void display() const override {
        cout << "Rectangle - ";
        Shape::display();
        cout << "Dimensions: " << width << "x" << height << ", Area: " << area() << endl;
    }
};

// Class with composition
class Engine {
private:
    int horsepower;
    
public:
    Engine(int hp) : horsepower(hp) {}
    
    void start() {
        cout << "Engine started (" << horsepower << " HP)" << endl;
    }
};

class Car {
private:
    string brand;
    Engine engine;
    
public:
    Car(string b, int hp) : brand(b), engine(hp) {}
    
    void start() {
        cout << brand << " car: ";
        engine.start();
    }
};

// Template class
template<typename T>
class Box {
private:
    T value;
    
public:
    Box(T v) : value(v) {}
    
    T getValue() const {
        return value;
    }
    
    void setValue(T v) {
        value = v;
    }
    
    void display() const {
        cout << "Box contains: " << value << endl;
    }
};

// Static members
class Counter {
private:
    static int count;
    int id;
    
public:
    Counter() {
        count++;
        id = count;
    }
    
    static int getCount() {
        return count;
    }
    
    int getId() const {
        return id;
    }
};

int Counter::count = 0;

int main() {
    // Polymorphism
    cout << "=== Polymorphism ===" << endl;
    Shape* shapes[3];
    shapes[0] = new Circle("Red", 5.0);
    shapes[1] = new Rectangle("Blue", 4.0, 6.0);
    shapes[2] = new Circle("Green", 3.0);
    
    for (int i = 0; i < 3; i++) {
        shapes[i]->display();
        cout << endl;
    }
    
    // Cleanup
    for (int i = 0; i < 3; i++) {
        delete shapes[i];
    }
    
    // Composition
    cout << "=== Composition ===" << endl;
    Car car1("Tesla", 450);
    Car car2("BMW", 350);
    
    car1.start();
    car2.start();
    
    // Templates
    cout << "\n=== Templates ===" << endl;
    Box<int> intBox(100);
    Box<string> strBox("Hello");
    Box<double> doubleBox(99.5);
    
    intBox.display();
    strBox.display();
    doubleBox.display();
    
    // Static members
    cout << "\n=== Static Members ===" << endl;
    Counter c1, c2, c3;
    cout << "Counter 1 ID: " << c1.getId() << endl;
    cout << "Counter 2 ID: " << c2.getId() << endl;
    cout << "Counter 3 ID: " << c3.getId() << endl;
    cout << "Total counters created: " << Counter::getCount() << endl;
    
    return 0;
}
